lib/import-notebook.js

```diff
@@ -47,89 +69,175 @@ export function importNotebook(event?: atom$CustomEvent) {
+ * also load the previous results and display them.
+ *
+ * @param {String} filename - Path of the file.
+ * @param {Boolean} importResults - Decides whether to display previous results
+ */
+export async function _loadNotebook(
+  filename: string,
+  importResults: boolean = false
+) {
   let data;
   let nb;
   try {
-    data = await readFileP(filename);
-    nb = parseNotebook(data);
+    data = JSON.parse(await readFileP(filename));
+    if (data.nbformat < 3) {
+      atom.notifications.addError("Only notebook version 4 is fully supported");
+      return;
+    } else if (data.nbformat == 3) {
+      atom.notifications.addWarning(
+        "Only notebook version 4 is fully supported"
+      );
+    }
+    nb = fromJS(data);
   } catch (err) {
     if (err.name === "SyntaxError") {
       atom.notifications.addError("Error not a valid notebook", {
         detail: err.stack
       });
     } else {
       atom.notifications.addError("Error reading file", {
-        detail: err.message
+        detail: err
       });
     }
     return;
   }
-  if (nb.nbformat < 4) {
-    atom.notifications.addError("Only notebook version 4 currently supported");
-    return;
-  }
   const editor = await atom.workspace.open();
   const grammar = getGrammarForNotebook(nb);
   if (!grammar) return;
-  editor.setGrammar(grammar);
+  atom.grammars.assignLanguageMode(editor, grammar.scopeName);
   const commentStartString = getCommentStartString(editor);
   if (!commentStartString) {
     atom.notifications.addError("No comment symbol defined in root scope");
     return;
   }
-  const sources = _.map(nb.cells, cell => {
-    return getCellSource(cell, commentStartString + " ");
+  const nbCells = [];
+  const sources = [];
+  const resultRows = [];
+  let previousBreakpoint = -1;
+
+  nb.cellOrder.forEach(value => {
+    const cell = nb.cellMap.get(value).toJS();
+    nbCells.push(cell);
+    const hyCell = toHydrogenCodeBlock(cell, commentStartString + " ");
+    resultRows.push(previousBreakpoint + hyCell.code.trim().split("\n").length);
+    previousBreakpoint += hyCell.row;
+    sources.push(hyCell.code);
   });
   editor.setText(sources.join(linesep));
+  if (importResults) importNotebookResults(editor, nbCells, resultRows);
 }
 
+/**
+ * Tries to determine the Atom Grammar of a notebook. Default is Python.
+ *
+ * @param {Notebook} nb - The Notebook to determine the Atom Grammar of.
+ * @return {atom$Grammar} - The grammar of the notebook.
+ */
 function getGrammarForNotebook(nb: Notebook) {
-  if (!nb.metadata.kernelspec || !nb.metadata.language_info) {
+  const {
+    kernelspec, // Often used in nbformat v4
+    kernel_info, // Official nbformat v4
+    language, // Sometimes used in nbformat v3
+    language_info // Offical nbformat v4
+  } = nb.metadata.toJS();
+  const kernel = kernelspec ? kernelspec : kernel_info;
+  const lang = language_info
+    ? language_info
+    : language
+    ? { name: language }
+    : null;
```

Comment:
```js
const lang = language_info ?
  language_info :
  language ?
    { name: language } :
    null
```

would look better and be consistent with the other parts of our code

Reply:
Agreed that was probably an auto lint as that looks nothing like i would do.